bin/runtime-common/src/messages_xcm_extension.rs

```diff
@@ -144,155 +141,199 @@ impl SenderAndLane {
+			enqueued_messages,
+		);
 	}
 }
 
-/// A structure that implements [`frame_support::traits::ProcessMessage`] and may
-/// be used in the `pallet-message-queue` configuration to stop processing messages when the
-/// bridge queue is congested.
+/// Manager of local XCM queues (and indirectly - underlying transport channels) that
+/// controls the queue state.
 ///
 /// It needs to be used at the source bridge hub.
-pub struct LocalXcmQueueMessageProcessor<Origin, Inner, R, MI, SL>(
-	PhantomData<(Origin, Inner, R, MI, SL)>,
-);
+pub struct LocalXcmQueueManager<H>(PhantomData<H>);
 
-impl<Origin, Inner, R, MI, SL> ProcessMessage
-	for LocalXcmQueueMessageProcessor<Origin, Inner, R, MI, SL>
-where
-	Origin: Clone
-		+ Into<MultiLocation>
-		+ FullCodec
-		+ MaxEncodedLen
-		+ Clone
-		+ Eq
-		+ PartialEq
-		+ TypeInfo
-		+ Debug,
-	Inner: ProcessMessage<Origin = Origin>,
-	R: MessagesConfig<MI>,
-	MI: 'static,
-	SL: Get<SenderAndLane>,
-{
-	type Origin = Origin;
-
-	fn process_message(
-		message: &[u8],
-		origin: Self::Origin,
-		meter: &mut WeightMeter,
-		id: &mut [u8; 32],
-	) -> Result<bool, ProcessMessageError> {
-		// if the queue is suspended, yield immediately
-		let sender_and_lane = SL::get();
-		let is_expected_origin = origin.clone().into() == sender_and_lane.location;
-		if is_expected_origin &&
-			LocalXcmQueueManager::is_inbound_queue_suspended::<R, MI>(sender_and_lane.lane)
-		{
-			return Err(ProcessMessageError::Yield)
+/// Maximal number of messages in the outbound bridge queue. Once we reach this limit, we
+/// send a "congestion" XCM message to the sending chain.
+const OUTBOUND_LANE_CONGESTED_THRESHOLD: MessageNonce = 8_192;
+
+/// After we have sent "congestion" XCM message to the sending chain, we wait until number
+/// of messages in the outbound bridge queue drops to this count, before sending `uncongestion`
+/// XCM message.
+const OUTBOUND_LANE_UNCONGESTED_THRESHOLD: MessageNonce = 1_024;
+
+impl<H: XcmBlobHauler> LocalXcmQueueManager<H> {
+	/// Must be called whenever we push a message to the bridge lane.
+	pub fn on_bridge_message_enqueued(
+		sender_and_lane: &SenderAndLane,
+		enqueued_messages: MessageNonce,
+	) {
+		// if we have already sent the congestion signal, we don't want to do anything
+		if Self::is_congested_signal_sent(sender_and_lane.lane) {
+			return
 		}
 
-		// else pass message to backed processor
-		Inner::process_message(message, origin, meter, id)
+		// if the bridge queue is not congested, we don't want to do anything
+		let is_congested = enqueued_messages > OUTBOUND_LANE_CONGESTED_THRESHOLD;
+		if !is_congested {
+			return
+		}
+
+		log::info!(
+			target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+			"Sending 'congested' XCM message to {:?} to avoid overloading lane {:?}: there are\
+			{} messages queued at the bridge queue",
+			sender_and_lane.location,
+			sender_and_lane.lane,
+			enqueued_messages,
+		);
+
+		if let Err(e) = Self::send_congested_signal(sender_and_lane) {
```

Comment:
maybe also to trigger some event here, I know that inner `send_xcm` creates one,
but maybe could be useful for later analyse if something happens